File: api-interop-layer/data/products/afd/AFDParser.js
/* eslint-disable no-plusplus */
const SPECIAL_HEADER_TYPES = {
  wwa: {
    re: /^\.(?<header>[A-Z]{3}\sWATCHES\/WARNINGS\/ADVISORIES)[.]{3}/,
  },
  tempsTable: {
    re: /^\.(?<header>.*TEMPS\/POPS)[.]{3}/,
  }
};

const GENERIC_HEADER_REGEX = /^\.(?<header>.+)[.]{3}/;
const SUBHEADER_REGEX = /^\s*\.{3}(?<subheader>[^.]+)\.{3}/;


export default class AFDParser {
  constructor(str){
    this.source = str;
    this.parsedNodes = [];
    this.currentContentType = "generic";

    // Bound methods
    this.parse = this.parse.bind(this);
    this.parseDocumentPreamble = this.parseDocumentPreamble.bind(this);
    this.parseSection = this.parseSection.bind(this);
    this.parseHeader = this.parseHeader.bind(this);
    this.parseSubheader = this.parseSubheader.bind(this);
    this.parseTextContent = this.parseTextContent.bind(this);
    this.parseWWAContent = this.parseWWAContent.bind(this);
    this.parseTempsTableContent = this.parseTempsTableContent.bind(this);
    this.parseEpilogueContent = this.parseEpilogueContent.bind(this);
    this.getStructureForTwig = this.getStructureForTwig.bind(this);
  }

  parse(){
    this.parsedNodes = [];
    // Parse out any preamble nodes, and set the
    // body to be everything _after_ the preamble
    let body = this.parseDocumentPreamble();

    // Stash the epilogue text for later, and set the body
    // to everything before it.
    const split = body.split(/\$\$/);
    body = split[0];
    const epilogueText = split[1];

    // Parse each of the sections.
    const sections = this.constructor.splitIntoTopicSections(body);
    sections.forEach(this.parseSection);

    // Now parse the epilogue text
    this.parseEpilogueContent(epilogueText);
  }

  /**
   * Parse out the preamble nodes,
   * and return a new string with the
   * corresponding section removed
   */
  parseDocumentPreamble(){
    const lines = this.source.split("\n");
    const preambleLines = [];
    let currentLine = 0;
    while(!lines[currentLine].match(/^\.[A-Za-z]/)){
      preambleLines.push(lines[currentLine]);
      currentLine += 1;
    }

    // Keep track of the string _after_ the preamble,
    // in order to return it to the caller
    const remaining = lines.slice(currentLine).join("\n");

    // Parse and push the preamble nodes based on the content
    const preambleParagraphs = preambleLines.join("\n").split("\n\n");
    for(let i = 0; i < preambleParagraphs.length; i++){
      let nodeType = "preambleText";
      if(i === 0){
        nodeType = "preambleCode";
      }
      const paragraph = preambleParagraphs[i];

      const text = this.constructor.normalizeSpaces(paragraph).trim();
      if(text !== ""){
        this.parsedNodes.push({
          type: nodeType,
          content: text
        });
      }
    }

    return remaining;
  }

  /**
   * Parse an AFD topic section.
   * The directive defines the topic sections as starting
   * with any of the valid headers (see header functions)
   * followed by any body text _up until_ double-ampersand.
   * We parse headers and body text for a single section below
   * using mapped functions for each section type as needed
   */
  parseSection(str){
    // Every new section is assumed to be a generic
    // type, until a special header dictates otherwise
    this.contentType = "generic";

    // Split into paragraphs, which are defined as
    // double newlines
    // If there is not a valid header in the first paragraph,
    // we consider it "empty"
    const paragraphs = str.trim().split("\n\n");
    paragraphs.forEach(paragraph => {
      let rest = this.parseHeader(paragraph);
      rest = this.parseSubheader(rest);
      this.parseTextContent(rest);
    });
  }

  parseHeader(str){
    // First, we see if we match on any of the more specialized
    // headers. If so, we append the corresponding node and update
    // the content type for the parser
    const specialTypeNames = Object.keys(SPECIAL_HEADER_TYPES);
    for(let i = 0; i < specialTypeNames.length; i++){
      const specialTypeName = specialTypeNames[i];
      const specialHeader = SPECIAL_HEADER_TYPES[specialTypeName];
      const match = str.match(specialHeader.re);
      if(match){
        let content = match.groups.header;
        if(specialTypeName === "wwa"){
          content = match.groups.header.replaceAll("/", "&hairsp;/&hairsp;");
        }
        this.contentType = specialTypeName;
        this.parsedNodes.push({
          type: "header",
          content
        });
        return str.slice(
          match.index + match[0].length
        );
      }
    }

    // If we get here, no specialized headers were found for this
    // paragraph. Attempt to parse a generic header
    const genericMatch = str.match(GENERIC_HEADER_REGEX);
    if(genericMatch){
      this.contentType = "generic";
      this.parsedNodes.push({
        type: "header",
        content: genericMatch.groups.header
      });
      return str.slice(
        genericMatch.index + genericMatch[0].length
      );
    }

    return str;
  }

  parseSubheader(str){
    const match = str.trim().match(SUBHEADER_REGEX);
    if(match){
      this.parsedNodes.push({
        type: "subheader",
        content: match.groups.subheader
      });
      
      return match.input.slice(
        match.index + match[0].length
      ).trim();
    }

    return str;
  }

  parseTextContent(str){
    if(str === ""){
      return str;
    }
    switch(this.contentType){
    case "wwa":
      return this.parseWWAContent(str);
    case "tempsTable":
      return this.parseTempsTableContent(str);
    default:
      return this.parsedNodes.push({
        type: "text",
        content: this.constructor.normalizeSpaces(
          str.trim().replaceAll("\n", " "))
      });
    }
    
  }

  parseWWAContent(str){
    // A newline followed by 4 spaces indentation
    // indicates line continuation. Replate with the
    // empty string, followed by a normal newline
    let currentString = str.trim();
    currentString = currentString.replace(/\n\s+/g, "\n");
    if(currentString !== ""){
      this.parsedNodes.push({
        type: "text",
        content: currentString
      });
    }
  }

  parseTempsTableContent(str){
    // There could be whitespace between the TEMPS/POPS header
    // and the actual data, so clean that up first. There
    // *shouldn't* be, but we've seen it happen,
    // so guard against it
    const lines = str.trim().split("\n");

    // We want to only consider lines that are part of table
    // rows. Anything else subsequently (if there is anything else)
    // should be considered text content.
    const tableRowRx = /[^\d]+(\d+\s+)+\/\s+(\d+\s+)+\d+\s*(\n|$)/g;
    const tableLines = lines.filter(line => line.match(tableRowRx));

    const restOfLines = lines.filter(line => !line.match(tableRowRx));

    // Parse out the row labels and the numbers from
    // any table rows
    const rx = /^([^\d]|\s)*/g;
    const rows = [];
    for(let i = 0; i < tableLines.length; i++){
      const line = tableLines[i];
      let numbers = line.split(rx).filter(item => (item !== "" || !item.match(/\s+/))).pop();

      numbers = numbers.trim().split(" ")
        .map(digitString => digitString.trim())
        .filter(digitString => digitString.match(/\d+/));

      const placeRx = /^(?<place>[^\d]+)/;
      let place = null;
      const placeMatch = line.match(placeRx);
      if(placeMatch){
        place = placeMatch.groups.place.trim();
      }

      rows.push({
        type: "temps-table-row",
        numbers,
        name: place
      });
    }

    if(rows.length){
      this.parsedNodes.push({
        type: "temps-table",
        rows
      });
    }

    // Finally, parse out any remaining text as
    // text node(s)
    const remainingText = restOfLines.join("\n").trim();
    if(remainingText !== ""){
      this.parsedNodes.push({
        type: "text",
        content: remainingText
      });
    }
  }

  parseEpilogueContent(str){
    const lines = str.trim().split("\n").map(line => line.trim());
    const currentString = lines.join("\n");
    if(currentString !== ""){
      this.parsedNodes.push({
        type: "epilogueText",
        content: currentString
      });
    }
  }

  getStructureForTwig(){
    const preambleCode = [];
    const preambleText = [];
    const body = [];
    const epilogue = [];
    this.parsedNodes.forEach(node => {
      if(node.type === "preambleCode"){
        preambleCode.push(node);
      } else if(node.type === "preambleText"){
        preambleText.push(node);
      } else if(node.type.startsWith("epilogue")){
        epilogue.push(node);
      } else {
        body.push(node);
      }
    });

    return {
      preamble: {
        code: preambleCode,
        text: preambleText
      },
      body,
      epilogue
    };
  }
  
  /**
   * Change all instances of multiple contiguous spaces
   * into a single space
   */
  static normalizeSpaces(str){
    return str.replace(/( +)|( $)/g, " ");
  }

  /**
   * Divide the given AFD string into "topic sections".
   * These are defined by the directive as beginning with a
   * header (see format above) and ending with "&&\n".
   * Note that we are a bit more forgiving here -- we will
   * simply split by the double-ampersand and assume these
   * are the topic sections, since WFOs sometimes forget headers
   */
  static splitIntoTopicSections(str){
    const splits = str.trimStart().split("&&\n");
    return splits.filter(section => !section.match(/^\s*$/));
  }
}

File: api-interop-layer/pkg/weather/data/afd_parser.go
package data

import (
	"regexp"
	"strings"
)

// AFDNode represents a parsed node in the AFD structure
type AFDNode struct {
	Type    string        `json:"type"`
	Content string        `json:"content,omitempty"`
	Rows    []AFDTempsRow `json:"rows,omitempty"`
}

type AFDTempsRow struct {
	Type    string   `json:"type"`
	Numbers []string `json:"numbers"`
	Name    string   `json:"name,omitempty"`
}

type AFDStructure struct {
	Preamble struct {
		Code []AFDNode `json:"code"`
		Text []AFDNode `json:"text"`
	} `json:"preamble"`
	Body     []AFDNode `json:"body"`
	Epilogue []AFDNode `json:"epilogue"`
}

var (
	// Regexes
	// TS: ^\.(?<header>[A-Z]{3}\sWATCHES\/WARNINGS\/ADVISORIES)[.]{3}
	// Go: (?P<header>...)
	reWWA = regexp.MustCompile(`^\.(?P<header>[A-Z]{3}\sWATCHES/WARNINGS/ADVISORIES)[.]{3}`)

	// TS: ^\.(?<header>.*TEMPS\/POPS)[.]{3}
	reTempsTable = regexp.MustCompile(`^\.(?P<header>.*TEMPS/POPS)[.]{3}`)

	// TS: ^\.(?<header>.+)[.]{3}
	reGenericHeader = regexp.MustCompile(`^\.(?P<header>.+)[.]{3}`)

	// TS: ^\s*\.{3}(?<subheader>[^.]+)\.{3}
	reSubheader = regexp.MustCompile(`^\s*\.{3}(?P<subheader>[^.]+)\.{3}`)

	// TS: ^\.[A-Za-z]
	rePreambleEnd = regexp.MustCompile(`^\.[A-Za-z]`)
)

type AFDParser struct {
	Source             string
	ParsedNodes        []AFDNode
	CurrentContentType string
}

func NewAFDParser(source string) *AFDParser {
	return &AFDParser{
		Source:             source,
		ParsedNodes:        []AFDNode{},
		CurrentContentType: "generic",
	}
}

func (p *AFDParser) Parse() {
	p.ParsedNodes = []AFDNode{}

	body := p.parseDocumentPreamble()

	// Split epilogue
	// TS: body.split(/\$\$/)
	parts := regexp.MustCompile(`\$\$`).Split(body, 2)
	bodyText := parts[0]
	var epilogueText string
	if len(parts) > 1 {
		epilogueText = parts[1]
	}

	// Sections
	// TS: splitIntoTopicSections -> split("&&\n")
	sections := p.splitIntoTopicSections(bodyText)
	for _, s := range sections {
		p.parseSection(s)
	}

	// Epilogue
	p.parseEpilogueContent(epilogueText)
}

func (p *AFDParser) GetStructureForTwig() AFDStructure {
	s := AFDStructure{
		Body:     []AFDNode{},
		Epilogue: []AFDNode{},
	}
	s.Preamble.Code = []AFDNode{}
	s.Preamble.Text = []AFDNode{}

	for _, node := range p.ParsedNodes {
		switch {
		case node.Type == "preambleCode":
			s.Preamble.Code = append(s.Preamble.Code, node)
		case node.Type == "preambleText":
			s.Preamble.Text = append(s.Preamble.Text, node)
		case strings.HasPrefix(node.Type, "epilogue"):
			s.Epilogue = append(s.Epilogue, node)
		default:
			s.Body = append(s.Body, node)
		}
	}
	return s
}

func (p *AFDParser) parseDocumentPreamble() string {
	lines := strings.Split(p.Source, "\n")
	preambleLines := []string{}
	currentLine := 0

	for currentLine < len(lines) {
		if rePreambleEnd.MatchString(lines[currentLine]) {
			break
		}
		preambleLines = append(preambleLines, lines[currentLine])
		currentLine++
	}

	remaining := ""
	if currentLine < len(lines) {
		remaining = strings.Join(lines[currentLine:], "\n")
	}

	preambleText := strings.Join(preambleLines, "\n")
	paragraphs := strings.Split(preambleText, "\n\n") // JS splits strictly? Go Split sep literal.

	for i, paragraph := range paragraphs {
		nodeType := "preambleText"
		if i == 0 {
			nodeType = "preambleCode"
		}
		text := normalizeSpaces(paragraph)
		text = strings.TrimSpace(text)

		if text != "" {
			p.ParsedNodes = append(p.ParsedNodes, AFDNode{
				Type:    nodeType,
				Content: text,
			})
		}
	}

	return remaining
}

func (p *AFDParser) parseSection(str string) {
	p.CurrentContentType = "generic"
	paragraphs := strings.Split(strings.TrimSpace(str), "\n\n")

	for _, paragraph := range paragraphs {
		rest := p.parseHeader(paragraph)
		rest = p.parseSubheader(rest)
		p.parseTextContent(rest)
	}
}

func (p *AFDParser) parseHeader(str string) string {
	// Special Headers
	// WWA
	if loc := reWWA.FindStringIndex(str); loc != nil {
		match := reWWA.FindStringSubmatch(str)
		// Header group is index 1
		header := match[1]
		header = strings.ReplaceAll(header, "/", "&hairsp;/&hairsp;") // WWA logic

		p.CurrentContentType = "wwa"
		p.ParsedNodes = append(p.ParsedNodes, AFDNode{
			Type:    "header",
			Content: header,
		})
		return str[loc[1]:] // Slice after match
	}

	// Temps Table
	if loc := reTempsTable.FindStringIndex(str); loc != nil {
		match := reTempsTable.FindStringSubmatch(str)
		header := match[1]
		p.CurrentContentType = "tempsTable"
		p.ParsedNodes = append(p.ParsedNodes, AFDNode{
			Type:    "header",
			Content: header,
		})
		return str[loc[1]:]
	}

	// Generic
	if loc := reGenericHeader.FindStringIndex(str); loc != nil {
		match := reGenericHeader.FindStringSubmatch(str)
		header := match[1]
		p.CurrentContentType = "generic"
		p.ParsedNodes = append(p.ParsedNodes, AFDNode{
			Type:    "header",
			Content: header,
		})
		return str[loc[1]:]
	}

	return str
}

func (p *AFDParser) parseSubheader(str string) string {
	strTrim := strings.TrimSpace(str)
	if loc := reSubheader.FindStringIndex(strTrim); loc != nil {
		match := reSubheader.FindStringSubmatch(strTrim)
		sub := match[1]
		p.ParsedNodes = append(p.ParsedNodes, AFDNode{
			Type:    "subheader",
			Content: sub,
		})

		// Wait, loc is relative to strTrim. I need to slice original str?
		// TS: match = str.trim().match(...)
		// return match.input.slice(...)
		// But Go doesn't give match.input easily unless I reused strTrim.
		// If I slice strTrim, it works.
		return strTrim[loc[1]:]
	}
	return str
}

func (p *AFDParser) parseTextContent(str string) {
	if str == "" {
		return
	}

	switch p.CurrentContentType {
	case "wwa":
		p.parseWWAContent(str)
	case "tempsTable":
		p.parseTempsTableContent(str)
	default:
		content := normalizeSpaces(strings.ReplaceAll(strings.TrimSpace(str), "\n", " "))
		p.ParsedNodes = append(p.ParsedNodes, AFDNode{
			Type:    "text",
			Content: content,
		})
	}
}

func (p *AFDParser) parseWWAContent(str string) {
	currentString := strings.TrimSpace(str)
	// replace(/\n\s+/g, "\n")
	reLines := regexp.MustCompile(`\n\s+`)
	currentString = reLines.ReplaceAllString(currentString, "\n")

	if currentString != "" {
		p.ParsedNodes = append(p.ParsedNodes, AFDNode{
			Type:    "text",
			Content: currentString,
		})
	}
}

func (p *AFDParser) parseTempsTableContent(str string) {
	lines := strings.Split(strings.TrimSpace(str), "\n")

	// TS: /[^\d]+(\d+\s+)+\/\s+(\d+\s+)+\d+\s*(\n|$)/g
	// Go doesn't support \s as reliably in all modes? \s is [ \t\n\f\r] usually.
	reTableRow := regexp.MustCompile(`[^\d]+(\d+\s+)+/\s+(\d+\s+)+\d+\s*(\n|$)`)

	tableLines := []string{}
	restOfLines := []string{}

	for _, line := range lines {
		if reTableRow.MatchString(line + "\n") { // Regex expects \n or end. Line doesn't have \n. Added it for match?
			// Actually match string doesn't need \n if it matches $
			if reTableRow.MatchString(line) {
				tableLines = append(tableLines, line)
			} else {
				restOfLines = append(restOfLines, line)
			}
		} else {
			// Redundant check?
			// Need to check exact match logic.
			// TS `line.match(rx)`.
			if reTableRow.MatchString(line) {
				tableLines = append(tableLines, line)
			} else {
				restOfLines = append(restOfLines, line)
			}
		}
	}

	// Process Rows
	rows := []AFDTempsRow{}
	reSplit := regexp.MustCompile(`^([^\d]|\s)*`) // split rx
	reDigits := regexp.MustCompile(`\d+`)
	rePlace := regexp.MustCompile(`^(?P<place>[^\d]+)`)

	for _, line := range tableLines {
		// TS: line.split(rx) -> ... -> numbers
		// Split by leading non-digits?
		// Actually TS logic: `split(rx).filter(...).pop()`
		// `rx` matches the prefix (place name).
		// So split gives ["", "numbers..."].
		parts := reSplit.Split(line, -1)
		// filter items
		validParts := []string{}
		for _, part := range parts {
			// TS: item !== "" || !item.match(/\s+/) ?
			// Actually TS: item !== "" || !item.match(/^\s+$/) check?
			// Usually split results in the parts separated.
			if strings.TrimSpace(part) != "" {
				validParts = append(validParts, part)
			}
		}

		// pop? The last part should be the numbers.
		var numbersStr string
		if len(validParts) > 0 {
			numbersStr = validParts[len(validParts)-1]
		}

		// Parse numbers
		nums := []string{}
		fields := strings.Fields(numbersStr)
		for _, f := range fields {
			if reDigits.MatchString(f) {
				nums = append(nums, f)
			}
		}

		// Place name
		var placeName string
		if match := rePlace.FindStringSubmatch(line); match != nil {
			placeName = strings.TrimSpace(match[1]) // group 1 is place
		}

		rows = append(rows, AFDTempsRow{
			Type:    "temps-table-row",
			Numbers: nums,
			Name:    placeName,
		})
	}

	if len(rows) > 0 {
		p.ParsedNodes = append(p.ParsedNodes, AFDNode{
			Type: "temps-table",
			Rows: rows,
		})
	}

	remainingText := strings.TrimSpace(strings.Join(restOfLines, "\n"))
	if remainingText != "" {
		p.ParsedNodes = append(p.ParsedNodes, AFDNode{
			Type:    "text",
			Content: remainingText,
		})
	}
}

func (p *AFDParser) parseEpilogueContent(str string) {
	lines := strings.Split(strings.TrimSpace(str), "\n")
	cleanLines := []string{}
	for _, l := range lines {
		cleanLines = append(cleanLines, strings.TrimSpace(l))
	}
	currentString := strings.Join(cleanLines, "\n")
	if currentString != "" {
		p.ParsedNodes = append(p.ParsedNodes, AFDNode{
			Type:    "epilogueText",
			Content: currentString,
		})
	}
}

func (p *AFDParser) splitIntoTopicSections(str string) []string {
	// TS: trimStart().split("&&\n")
	trimmed := strings.TrimLeft(str, " \t\n")
	splits := strings.Split(trimmed, "&&\n")

	valid := []string{}
	reSpace := regexp.MustCompile(`^\s*$`)
	for _, s := range splits {
		if !reSpace.MatchString(s) {
			valid = append(valid, s)
		}
	}
	return valid
}

func normalizeSpaces(str string) string {
	// TS: replace(/( +)|( $)/g, " ")
	re := regexp.MustCompile(`( +)|( $)`)
	return re.ReplaceAllString(str, " ")
}
